group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java

```diff
@@ -677,6 +705,73 @@ private void maybeRemovePartitionEpoch(
         }
     }
 
+    /**
+     * Removes the partition epochs based on the provided assignment.
+     *
+     * @param assignment    The assignment.
+     * @param expectedEpoch The expected epoch.
+     * @throws IllegalStateException if the epoch does not match the expected one.
+     * package-private for testing.
+     */
+    void removePartitionEpochs(
+        Map<Uuid, Set<Integer>> assignment,
+        int expectedEpoch
+    ) {
+        assignment.forEach((topicId, assignedPartitions) -> {
+            currentPartitionEpoch.compute(topicId, (__, partitionsOrNull) -> {
+                if (partitionsOrNull != null) {
+                    assignedPartitions.forEach(partitionId -> {
+                        Integer prevValue = partitionsOrNull.remove(partitionId);
+                        if (prevValue != expectedEpoch) {
+                            throw new IllegalStateException(
+                                String.format("Cannot remove the epoch %d from %s-%s because the partition is " +
+                                    "still owned at a different epoch %d", expectedEpoch, topicId, partitionId, prevValue));
+                        }
+                    });
+                    if (partitionsOrNull.isEmpty()) {
+                        return null;
+                    } else {
+                        return partitionsOrNull;
+                    }
+                } else {
+                    throw new IllegalStateException(
+                        String.format("Cannot remove the epoch %d from %s because it does not have any epoch",
+                            expectedEpoch, topicId));
+                }
+            });
+        });
+    }
+
+    /**
+     * Adds the partitions epoch based on the provided assignment.
+     *
+     * @param assignment    The assignment.
+     * @param epoch         The new epoch.
+     * @throws IllegalStateException if the partition already has an epoch assigned.
+     * package-private for testing.
+     */
+    void addPartitionEpochs(
+        Map<Uuid, Set<Integer>> assignment,
+        int epoch
+    ) {
+        assignment.forEach((topicId, assignedPartitions) -> {
+            currentPartitionEpoch.compute(topicId, (__, partitionsOrNull) -> {
+                if (partitionsOrNull == null) {
+                    partitionsOrNull = new TimelineHashMap<>(snapshotRegistry, assignedPartitions.size());
+                }
+                for (Integer partitionId : assignedPartitions) {
+                    Integer prevValue = partitionsOrNull.put(partitionId, epoch);
```

Comment:
We don't create new `partitionsOrNull` when updating `partitionsOrNull`, so should we use `putIfAbsent` to replace `put` to avoid overriding the epoch when the partition is still owned by another epoch?

for another, should we view `addPartitionEpochs`  as transaction op? that means we update the `currentPartitionEpoch` only if all elements in `assignment` are valid.

Reply:
BTW, I have the same question for `removePartitionEpochs`

Reply:
As per my understanding, `currentPartitionEpoch` is backed by `TimelineHashMap` where soft state is updated here but if exception occurs then the transaction is not persisted and gets overriden on next successful update. This was an existing code in ConsumerGroup.java which has been put back in current PR, we initially moved this code to ModernGroup so both Share and Consumer groups can use but later realized that Share Groups do not need this functionality.

Reply:
Sorry that I'm in traveling so I can't check it on my computer. The following description is my guess.

It seems to me partitionOrNull should not be updated by "epoch" if it already has an epoch assigned, right? If so, even though we throw exception here, the method "put" still changes the partitionOrNull.

Reply:
> If so, even though we throw exception here, the method "put" still changes the partitionOrNull.

Yes, that's right. But it's a soft update which gets upadted when CoordinatorRuntime either does rollback or commit. So I don't think it should concern us. Adding @dajac for more context.

Reply:
> Yes, that's right. But it's a soft update which gets upadted when CoordinatorRuntime either does rollback or commit. So I don't think it should concern us.

thanks for explanation. I overlook the `SnapshotRegistry` :(